feat: Auto convert between oiio:ColorSpace and CICP attributes in I/O

brechtvl opened the pull request:
## Description

When reading image files with CICP metadata, automatically set the corresponding "oiio:ColorSpace". When writing files that support CICP and no other colorspace metadata can represent "oiio:ColorSpace", automatically write CICP metadata.

Setting "oiio:ColorSpace" on read prefers scene referred over display referred color spaces, changing existing behavior as little as possible. The alternative would have been to interpret the presence of CICP metadata as an indication that the image is likely display referred, which might be reasonable too. Either way it's a guess.

There is no automatic mapping from `g22_rec709_display` to CICP currently to keep behavior unchanged, as this is often not what you want and further discussion is needed to decided on the right behavior.

Also add new ColorConfig `get_cicp` and `get_color_interop_id` API functions to share logic between file formats.

## Tests

Tests were updated, the auto detected color space name appears in the output.

Commands like these should also do the right thing automatically.
```
oiiotool in.exr --ociodisplay "Rec.2100-PQ - Display" "ACES 2.0 - HDR 1000 nits (P3 D65)" -o out.avif
oiiotool out.avif --autocc -o out.exr
```

## Checklist:

- [x] I have read the [contribution guidelines](https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/main/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable. (Check if there is no
  need to update the documentation, for example if this is a bug fix that
  doesn't change the API.)
- [x] I have ensured that the change is tested somewhere in the testsuite
  (adding new test cases if necessary).
- [x] If I added or modified a C++ API call, I have also amended the
  corresponding Python bindings (and if altering ImageBufAlgo functions, also
  exposed the new functionality as oiiotool options).
- [x] My code follows the prevailing code style of this project. If I haven't
  already run clang-format before submitting, I definitely will look at the CI
  test that runs clang-format and fix anything that it highlights as being
  nonconforming.

zachlewis commented:
Oh hells yes! Brecht, this is fantastic! I've been slowly working on my own p.o.c. (in python) for managing the relationship between CICP <--> colorInteropID <--> OCIO, and what you've put together aligns pretty closely with what I've been tinkering with. I really appreciate the initiative, this is exactly the direction we should be heading towards!


I have a few suggestions + ideas, for your consideration:


- Convert CICP values to and from an intermediate string `colorInteropID` attribute, instead of directly embedding a CIF interop ID token as a `oiio:ColorSpace` attribute. 
  - I think we should have a layer of abstraction between external non-OCIO color metadata attributes as wrangled by format readers and writers, and OIIO's internal OCIO-config-specific accounting of color and image state
  - `oiio:ColorSpace` is an internal (non-serialized) attribute whose value always refers to a specific OCIO config that OIIO knows about. 
  - `colorInteropID` is an external (serializable) attribute whose value reliably informs and reflects other representations of the data's image state / color encoding. It functions sort of like a proxy, or a bridge between different color-managed frameworks. 
- I think we should have dedicated functions for reconciling, interpreting, and translating between "external" non-OCIO-specific metadata constructs (e.g., `CICP`,  `oiio:InputFilePath`, `acesImageContainerFlag`, `MDCV`, etc...) and OCIO-config-specific metadata (primarily, `oiio:ColorSpace`), with `colorInteropID` serving as a means for translating between them, where appropriate. 
  - This would allow us to defer the `colorInteropID` <--> `oiio:ColorSpace` resolution as late as possible; while still allowing users to make meaningful edits to incoming or outgoing CICP (and MDCV) metadata.
  - It would also provide natural junctures for updating / synchronizing / managing the relationship between the internal accounting of the image state, and the external metadata we persist / blast / adjust).
    - This makes room for us to add more nuanced CICP (and MDCV) read/write behavior in the future without API/ABI-breaking changes
      - E.g., maybe users want all CICP-supporting format readers to immediately set the `colorInteropID` attribute for CICP data signalling HDR encodings, but would otherwise like to exhaust OCIO file rules first for SDR encodings, or otherwise manually override individual CICP values before applying a color-aware IBA. I'm thinking of users trying to troubleshoot format / encoding / application specific CICP behavior experienced outside of OIIO.

Damn, I have more to add, of course, but I've got a bus to catch!

zachlewis commented:
To be honest, I don't really have a problem with setting oiio:ColorSpace as early as possible; I just wonder if it's necessary, and if doing so imposes complexity of trying to keep CICP and oiio:ColorSpace attributes in sync. Either way, we need a mechanism for divining valid oiio:ColorSpace values just in time for color-aware IBAs that are otherwise missing and needing the attribute. 

My thinking is, if a user isn't using a color-aware IBA, it's probably safe to persist existing CICP primaries and trc data for as long as possible. But as soon as a color-aware IBA is encountered, incoming CICP data is potentially invalidated, so it may as well be removed. This would permit users to manually edit CICP metadata (in whole or in part) either before applying color-aware IBAs or just before writing to supported formats, allowing folks to pretty effortlessly preview the effects of modifying existing CICP data without having to worry about whether modifications to CICP metadata reflect the current state of the oiio:ColorSpace attribute, or vice versa. 

Conversely, once we're working with color-aware IBAs, we only care about keeping the value of oiio:ColorSpace up to date. If we blast away the CICP attribute every time we set oiio:ColorSpace, then when it comes time to setting CICP metadata for format writers, there's no ambiguity in what to do:
 - If CICP is set, it means either the buffer never encountered any IBAs that would affect color, or it means the user has manually set or edited the value intentionally -- either way, use as-is.
 - If CICP is unset and oiio:ColorSpace is set, use either the corresponding color space's interop_id attribute to inform the primaries and transfer function CICP values; or otherwise try to infer CICP values by attempting to match oiio:ColorSpace's definition to one of the canonical CIF interop_ids that unambiguously map to CICP pri + trc values. 
(Crucially, this would allow config authors to control, for example, whether a Gamma 2.2-based Display/View transform elicits a PNG tagged as 1-13-1)


Likewise, if we defer the automatic derivation of "oiio:ColorSpace" to the last possible moment, that means that if the value happens to be set, either the user has set it manually, OR it was set by an upstream IBA applied by the user. If the user is only interested in setting or modifying CICP data, I don't think we should create conditions that would potentially invalidate the validity of an oiio:ColorSpace attribute initially set on read. 

On the other hand, we certainly *could* endeavor to always update oiio:ColorSpace whenever CICP is updated. But given that color-aware IBAs might specify custom configs, I think it makes more sense to couple CICP mutations with corresponding changes to the config-agnostic "colorInteropID" attribute, interpreting the value as necessary for the given IBA / OCIO config. 


As for the behavior of the colorInteropID attribute itself, I think it should disappear and reappear as necessary, just like the CICP does, as described above. 

I'm not totally against allowing users to use CIF strings directly as oiio:ColorSpace values -- I think it would be a nice convenience, personally -- but I think it might be a slippery slope into compounded complexity down the road, when we want to use the colorInteropID to convey "namespaced" color spaces, which might require searching for a config, and then searching for a color space in that config, and then either finding an equivalent color space in the active config, or otherwise creating an ad-hoc transform / Processor into the current config. 

In other words, given the limited scope and use of the "unnamespaced" / reserved set of CIF strings, I think treating them as if they were known OCIO color spaces is more or less fine; but given the other applications of the colorInteropID attribute, we do need to treat it as a first-class citizen.

Does this make sense?

lgritz commented:
I was thinking that the reader is primarily responsible for setting the attribs that directly come from the file (like "CICP" from these files, or "colorInteropID" for exr, etc.). Optionally, it can also set "oiio:ColorSpace" if it knows for sure what it should be set to, like if the CICP exactly corresponds to a standard interop ID value that will be available in any default config, but this is not necessarily required.

Then, we want a global `deduce_color_space()` (or name TBD) that takes a spec and config and is tasked with setting "oiio:ColorSpace" (if not already set, or perhaps unconditionally if a flag parameter says so) based on... the full set of what's in the spec -- colorInteropID, CICD, filename, any other hints. This utility can embody the centralized logic that expresses our best ideas for how to adjudicate the possibly under- or over-specified or contradictory information that might be in the spec.

The few functions that really need to know what color space the data is in can call deduce_color_space() before doing any critical color operations.

I think that if we can figure out that there is an interopid that corresponds to whatever this image is, then oiio:ColorSpace should always be that. But I don't think it necessarily needs to be restricted to interop id names in cases where something a file supports isn't expressible as an OCIO color space name. In those cases, of course, we wouldn't expect an OCIO-based color transform to be able to deal with it, but we (OIIO) may internally handle other cases.

I do agree that oiio:ColorSpace is non-serialized -- it's never written or read from any file, it's just an internal marker for our best inference (based on what was in the file or other info) about what the data is.

brechtvl commented:
I changed the public API functions to be `getColorInteropID(cicp)` and `getCICP(colorspace)`. This way they don't include logic about how to resolve colorspaces from attributes, since it looks like that's going to change in the future. At the cost of a small amount of duplicated code for each file format.

Also added Python bindings for those.

lgritz commented:
OK, whatever you guys decide here is fine with me. The last several rounds of this discussion has been on esoteric points that are totally over my head. I trust you to make the right decisions, just let me know when you both agree that what we have here is ready to merge.

I've already got the Dec 1's scheduled 3.0 and 3.1 releases staged and ready to go, none of this is going to make that deadline, so whatever we merge here is going to get tried out in main for a few weeks before being backported to become a tagged release, so I think the stakes are low even if you merge something that you decide to revise reasonably soon after.

zachlewis commented:
> I've already got the Dec 1's scheduled 3.0 and 3.1 releases staged and ready to go, none of this is going to make that deadline, so whatever we merge here is going to get tried out in main for a few weeks before being backported to become a tagged release, so I think the stakes are low even if you merge something that you decide to revise reasonably soon after.

I think, in general, throughout the 3.1 release cycle, we should caveat these color management features as "experimental" and subject to change. I think there's a good chance we won't be able to effectively solicit enough meaningful feedback in some areas without a concrete implementation for folks to mess around with; and I know the OpenColorIO TSC is waiting to see what we end up doing here with stuff like CICP before providing more "official" guidance. Also, I expect that we'll make incremental improvements and incorporate additional features along the way that are outside the scope of what we're immediately trying to achieve. So, it would be cool if we could "reserve the right" to adjust behavior between OIIO patch versions. This would also serve to keep the relevant PRs less painful to review and more pleasurable to merge.

lgritz commented:
> I think, in general, throughout the 3.1 release cycle, we should caveat these color management features as "experimental" and subject to change. 

That sounds reasonable. 

Though I think that when we're talking about behavior, and not API/ABI compatibility that will cause downstream software to not build correctly, we're allowed to "fix bugs" even if it changes behavior by a reasonable amount, which gives us some leeway naturally to say "this behavior might be changing, but it was wrong before, and we're improving it" at a patch release.

Nevertheless, if you think this area is likely to have repeated churn and changes in direction (not just monotonically increasing the set of color scenarios we recognize and handle), then we can add a line to future release notes saying that we're expecting extra change.

brechtvl commented on src/libOpenImageIO/color_ocio.cpp:
```diff
@@ -1997,6 +1997,173 @@ ColorConfig::parseColorSpaceFromString(string_view str) const
+constexpr ColorInteropID color_interop_ids[] = {
+    // Scene referred interop IDs first so they are the default in automatic
+    // conversion from CICP to interop ID. Some are not display color spaces
+    // at all, but can be represented by CICP anyway.
```
Yes, as mentioned in the description I think that's reasonable too.

There are some potentially negative consequences:
* OIIO currently reads sRGB images as `srgb_rec709_scene`. If the presence of CICP would change that to `srgb_rec709_display`, some applications integrating OIIO would need to be updated (again) to handle both cases.
* If you use `srgb_p3d65_scene` for textures, then it would help to write CICP to make them open and display correctly in other software. But then opening them again in OIIO they would be `srgb_p3d65_display`.
* Other software for painting and editing textures may write CICP (maybe more so in the future if AVIF, JPEG XL gain more adoption).

lgritz replied:
>  we should leave this behavior as-is for the immediate future, and continue with implementing a more robust strategy in a forthcoming PR

Fine by me. We have two gates: one on individual PRs, but another on when we backport a batch of PRs to 3.1. It's ok to noodle with the specifics from PR to PR as long as we have things reasonably settled down for the backporting steps.

brechtvl commented:
> There are several other PRs in flight now that are drafts because they depend on this being merged first. So I want to put a premium on expeditiously getting this first one acceptable to merge, even if we need to come back and fine tune afterwards.

For this PR, the tricky decisions are how to write `g22_rec709_display` and how to guess scene referred or display referred from CICP. In both cases I've left the existing behavior unchanged (to the extent possible), to avoid approval being blocked by a decision for those.

zachlewis commented:
Understood; we can discuss and make changes to specifically address these
issues in subsequent issues / PRs. For the sake of this PR, I think this is
a fine place to leave things, especially if it helps make things easier to
align with what you’ve done with Blender.

I’m traveling tomorrow, but over the next couple of days, I’ll have a
closer look at the code ASAP —  absent anything glaring, I very much expect
to approve.



On Mon, Dec 1, 2025 at 6:53 PM Brecht Van Lommel ***@***.***>
wrote:

> *brechtvl* left a comment (AcademySoftwareFoundation/OpenImageIO#4964)
> <https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/4964#issuecomment-3599484752>
>
> There are several other PRs in flight now that are drafts because they
> depend on this being merged first. So I want to put a premium on
> expeditiously getting this first one acceptable to merge, even if we need
> to come back and fine tune afterwards.
>
> For this PR, the tricky decisions are how to write g22_rec709_display and
> how to guess scene referred or display referred from CICP. In both cases
> I've left the existing behavior unchanged, to avoid approval being blocked
> by a decision for those.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/4964#issuecomment-3599484752>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARAC4ADHVPIJHIRBSQU3BD37TIH5AVCNFSM6AAAAACNLYHACGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJZGQ4DINZVGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
